### PR TITLE
Kmean doc minor fixes

### LINKED ISSUE #3473: K-means documentation doesn't explain outputs' meanings
The only explanation of K-means outputs is:

Output Column Name | Column Type | Description
-- | -- | --
Score | Single | The unbounded score that was calculated by the trainer to determine the prediction.
PredictedLabel | Int32 | The cluster id predicted by the trainer.

I don't feel they are clear enough to users.

### LINKED ISSUE #3494: KMeansTrainer documentation doesn't reference the extension methods that create it
The remarks say "To create this trainer, use KMeansTrainer", with a link to the same page.
It should instead give two links to the two extension methods in `KMeansClusteringExtensions`.

### LINKED ISSUE #3514: KmeansTrainer sasys use itself to create itself
A trainer should be created by its public API, not itself.

image

This trainer's documentation also has formatting problems such as missing space.

### PR DESCRIPTION
**Pull request by wschin**
Fix #3473, fix #3514, and fix #3494.

### CODE REVIEW

File: docs/api-reference/io-columns-clustering.md [hunk 1]
```diff
@@ -3,5 +3,5 @@ The input features column data must be <xref:System.Single>. No label column nee
 
 | Output Column Name | Column Type | Description|
 | -- | -- | -- |
-| `Score` | <xref:System.Single> | The unbounded score that was calculated by the trainer to determine the prediction.|
-| `PredictedLabel` | <xref:System.Int32> | The cluster id predicted by the trainer.|
\ No newline at end of file
+| `Score` | <xref:System.Single> | The unbounded score that was calculated by the trainer to determine the prediction. |
```
Comment (natke) [inline docs/api-reference/io-columns-clustering.md]: Is `unbounded` the right word here? Would `raw` work equally well or better?

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 2]
```diff
@@ -44,13 +45,20 @@ namespace Microsoft.ML.Trainers
...
+    /// With K-means, the data is clustered into a specified number of clusters in order to minimize the within-cluster sum of squared distances.
     /// K-means++ improves upon K-means by using the [Yinyang K-Means](https://research.microsoft.com/apps/pubs/default.aspx?id=252149)
     /// method for choosing the initial cluster centers.
-    /// K-Means++ accelerates K-Means up to an order of magnitude while producing exactly the same clustering results(modulo floating point precision issues).
+    /// K-Means++ accelerates K-Means up to an order of magnitude while producing exactly the same clustering results (modulo floating point precision issues).
     /// K-Means++ observes that there is a lot of redundancy across iterations in the KMeans algorithms and most points do not change their clusters during an iteration.
     /// It uses various bounding techniques to identify this redundancy and eliminate many distance computations and optimize centroid computations.
+    ///
+    /// ### Scorning Function
+    /// The output Score column contains the L2-norm distance of the given input vector $\textbf{x}\in \mathbb{R}^n$ to each cluster's centroid.
```
Comment (natke) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: Score --> score

Comment (natke) [reply src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: Would saying Euclidean distance be clearer?

File: docs/api-reference/io-columns-clustering.md (hunk 1 above)
Comment (yaeldekel) [reply docs/api-reference/io-columns-clustering.md]: Isn't the Score column in clustering actually a vector of the distances from the cluster centers?

---
In reply to: [277476207](https://github.com/dotnet/machinelearning/pull/3520#discussion_r277476207) [](ancestors = 277476207)

Comment (yaeldekel) [inline docs/api-reference/io-columns-clustering.md]: >to determine the prediction [](start = 90, length = 27)

I would change to "to determine which cluster the example belongs to".

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 3]
```diff
@@ -44,13 +45,20 @@ namespace Microsoft.ML.Trainers
...
+    /// With K-means, the data is clustered into a specified number of clusters in order to minimize the within-cluster sum of squared distances.
     /// K-means++ improves upon K-means by using the [Yinyang K-Means](https://research.microsoft.com/apps/pubs/default.aspx?id=252149)
     /// method for choosing the initial cluster centers.
-    /// K-Means++ accelerates K-Means up to an order of magnitude while producing exactly the same clustering results(modulo floating point precision issues).
+    /// K-Means++ accelerates K-Means up to an order of magnitude while producing exactly the same clustering results (modulo floating point precision issues).
     /// K-Means++ observes that there is a lot of redundancy across iterations in the KMeans algorithms and most points do not change their clusters during an iteration.
     /// It uses various bounding techniques to identify this redundancy and eliminate many distance computations and optimize centroid computations.
+    ///
+    /// ### Scorning Function
+    /// The output score column contains the L2-norm distance of the given input vector $\textbf{x}\in \mathbb{R}^n$ to each cluster's centroid.
```
Comment (yaeldekel) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: >L2 [](start = 45, length = 2)

Should this be in LaTex as well?

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 4]
```diff
@@ -44,13 +45,20 @@ namespace Microsoft.ML.Trainers
...
+    /// Assume that the centriod of the $c$-th cluster is $\textbf{m}_c \mathbb{R}^n$.
```
Comment (yaeldekel) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: >  [](start = 71, length = 1)

Missing `\in`

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 5]
```diff
@@ -44,13 +45,20 @@ namespace Microsoft.ML.Trainers
...
+    /// The $c$-th value at the Score column would be $d_c = || \textbf{x} - \textbf{m}_c ||_2^2$.
+    /// The predicted label is the index with the smallest value in a $K$ dimension vector $[d_1, \dots, d_K]$, where $K$ is the number of clusters.
```
Comment (yaeldekel) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: >[d_1, \dots, d_K] [](start = 92, length = 17)

I would write the indices from 0 to K-1.

File: docs/api-reference/io-columns-clustering.md (hunk 1 above)
Comment (wschin) [reply docs/api-reference/io-columns-clustering.md]: Ok.

---
In reply to: [277750939](https://github.com/dotnet/machinelearning/pull/3520#discussion_r277750939) [](ancestors = 277750939)

Comment (wschin) [reply docs/api-reference/io-columns-clustering.md]: Will do `The distances of the given data point to all clusters' centriods.`

---
In reply to: [277750343](https://github.com/dotnet/machinelearning/pull/3520#discussion_r277750343) [](ancestors = 277750343,277476207)

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs (hunk 2 above)
Comment (wschin) [reply src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: I prefer equations over words but I will mention `Euclidean distance`.

---
In reply to: [277476454](https://github.com/dotnet/machinelearning/pull/3520#discussion_r277476454) [](ancestors = 277476454)

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs (hunk 5 above)
Comment (wschin) [reply src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: Ok.

---
In reply to: [277753639](https://github.com/dotnet/machinelearning/pull/3520#discussion_r277753639) [](ancestors = 277753639)

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs (hunk 3 above)
Comment (wschin) [reply src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: Ok. Will do $L_2$.

---
In reply to: [277752467](https://github.com/dotnet/machinelearning/pull/3520#discussion_r277752467) [](ancestors = 277752467)

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs (hunk 4 above)
Comment (wschin) [reply src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: Thanks!

---
In reply to: [277752959](https://github.com/dotnet/machinelearning/pull/3520#discussion_r277752959) [](ancestors = 277752959)

**Pull request merged**